state/protocol/badger/mutator.go

```diff
@@ -511,6 +517,110 @@ func (m *Mutator) sealExtend(candidate *flow.Block) (*flow.Seal, error) {
+// correspond to blocks on the fork, should not appear more than once on a fork,
+// and should be sorted by block height (within a payload).
+func (m *Mutator) receiptExtend(candidate *flow.Block) error {
+	blockID := candidate.ID()
+	m.state.tracer.StartSpan(blockID, trace.ProtoStateMutatorExtendCheckReceipts)
+	defer m.state.tracer.FinishSpan(blockID, trace.ProtoStateMutatorExtendCheckReceipts)
+
+	header := candidate.Header
+	payload := candidate.Payload
+
+	var sealedHeight uint64
+	err := m.state.db.View(operation.RetrieveSealedHeight(&sealedHeight))
+	if err != nil {
+		return fmt.Errorf("could no retrieve sealed height: %w", err)
+	}
+
+	// forkBlocks is used to keep the IDs of the blocks we iterate through. We
+	// use it to skip receipts that are not for blocks in the fork.
+	forkBlocks := make(map[flow.Identifier]struct{})
+
+	// Create a lookup table of all the receipts that are already in blocks.
+	// This will be used to filter out duplicates.
+	lookup := make(map[flow.Identifier]struct{})
+
+	// loop through the fork backwards, from parent to last sealed, and keep
+	// track of blocks and receipts visited on the way.
+	ancestorID := header.ParentID
+	for {
+
+		// keep track of blocks we iterate over
+		forkBlocks[ancestorID] = struct{}{}
+
+		ancestor, err := m.state.headers.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor header (%x): %w", ancestorID, err)
+		}
+
+		index, err := m.state.index.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor index (%x): %w", ancestorID, err)
+		}
+
+		// keep track of all receipts we iterate over
+		for _, recID := range index.ReceiptIDs {
+			lookup[recID] = struct{}{}
+		}
+
+		// break out when we reach the sealed height
+		if ancestor.Height <= sealedHeight {
+			break
+		}
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// prevReceiptHeight is used to check that receipts within a payload are
+	// sorted by block height. For each receipt in the payload, we check that
+	// its block height is greater or equal than the previous receipt.
+	var prevReceiptHeight uint64 = 0
+
+	// check each receipt included in the payload for duplication
+	for _, receipt := range payload.Receipts {
+
+		// if the receipt was already included before, error
+		_, duplicated := lookup[receipt.ID()]
+		if duplicated {
+			return state.NewInvalidExtensionErrorf("payload includes duplicate receipt (%x)", receipt.ID())
+		}
+
+		// if the receipt is not for a block on this fork, error
+		if _, ok := forkBlocks[receipt.ExecutionResult.BlockID]; !ok {
+			return state.NewInvalidExtensionErrorf("payload includes receipt for block not on fork (%x)", receipt.ExecutionResult.BlockID)
+		}
+
+		// get the reference block to check expiry
+		h, err := m.state.headers.ByBlockID(receipt.ExecutionResult.BlockID)
+		if err != nil {
+			return fmt.Errorf("could not get reference block (%x): %w", receipt.ExecutionResult.BlockID, err)
+		}
```

Comment:
This repeated lookup can be easily avoided by changing https://github.com/onflow/flow-go/blob/32812eafcd71f3fe44e8e95c8646f7ae2d74d50d/state/protocol/badger/mutator.go#L539 to 
```golang
forkBlocks := make(map[flow.Identifier]*flow.Header)
```